### PR TITLE
Add OPA support for task stages

### PR DESCRIPTION
**Pull request by mrinalirao**
## Description

Add OPA support to the go-tfe client for policy API's. OPA is still in beta phase.

This PR only contains changes to the[ Task Stage API](https://github.com/hashicorp/terraform-docs-common/pull/141/files#diff-c87624a2e9ea0583c4c08814f1fe930c72f385e30133d19b3085a817d416132a).

[API Docs](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/run-tasks/run-task-stages-and-results)

For a complete list of API's that were updated see : https://github.com/hashicorp/terraform-docs-common/pull/141/files

Added Api's:

-Task stage override

## Testing plan

Integration tests added

## External links

[API Docs](https://developer.hashicorp.com/terraform/cloud-docs/api-docs/run-tasks/run-task-stages-and-results)
[JIRA](https://hashicorp.atlassian.net/browse/TF-1450)
[API PR](https://github.com/hashicorp/terraform-docs-common/pull/141/files)

## Output from tests
Including output from tests may require access to a TFE instance. Ignore this section if you have no environment to test against.

$ envchain local go test -run TestTaskStagesRead -v ./... -tags=integration

```
=== RUN   TestTaskStagesRead_Beta
--- PASS: TestTaskStagesRead_Beta (35.77s)
=== RUN   TestTaskStagesRead_Beta/without_read_options
    --- PASS: TestTaskStagesRead_Beta/without_read_options (0.20s)
=== RUN   TestTaskStagesRead_Beta/with_include_param_task_results
    --- PASS: TestTaskStagesRead_Beta/with_include_param_task_results (0.21s)
=== RUN   TestTaskStagesRead_Beta/with_include_param_task_results/task_results_are_properly_decoded
        --- PASS: TestTaskStagesRead_Beta/with_include_param_task_results/task_results_are_properly_decoded (0.00s)
=== RUN   TestTaskStagesRead_Beta/with_include_param_task_results/policy_evaluations_are_properly_decoded
        --- PASS: TestTaskStagesRead_Beta/with_include_param_task_results/policy_evaluations_are_properly_decoded (0.00s)
PASS
ok      github.com/hashicorp/go-tfe     19.201s
?       github.com/hashicorp/go-tfe/examples/organizations      [no test files]
?       github.com/hashicorp/go-tfe/examples/workspaces [no test files]
?       github.com/hashicorp/go-tfe/mocks       [no test files]
```

$ envchain local go test -run TestTaskStagesList_Beta -v ./... -tags=integration

```
=== RUN   TestTaskStagesList_Beta
--- PASS: TestTaskStagesList_Beta (25.35s)
=== RUN   TestTaskStagesList_Beta/with_no_params
    --- PASS: TestTaskStagesList_Beta/with_no_params (0.23s)
PASS
ok      github.com/hashicorp/go-tfe     26.375s
?       github.com/hashicorp/go-tfe/examples/organizations      [no test files]
?       github.com/hashicorp/go-tfe/examples/workspaces [no test files]
?       github.com/hashicorp/go-tfe/mocks       [no test files]
```

$ envchain local go test -run TestTaskStageOverride_Beta -v ./... -tags=integration

```
=== RUN   TestTaskStageOverride_Beta
--- PASS: TestTaskStageOverride_Beta (93.51s)
=== RUN   TestTaskStageOverride_Beta/when_the_policy_failed
    helper_test.go:1037: Polling run "run-ANbRt1ddSdRTCahK" for status included in ["post_plan_awaiting_decision" "errored"] with deadline of 2022-11-10 14:39:28.039298 +1100 AEDT m=+141.946231751
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "plan_queued"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "planning"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "planning"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-ANbRt1ddSdRTCahK"
    helper_test.go:1049: Run "run-ANbRt1ddSdRTCahK" had status "post_plan_awaiting_decision"
    --- PASS: TestTaskStageOverride_Beta/when_the_policy_failed (41.02s)
=== RUN   TestTaskStageOverride_Beta/when_the_policy_passed
    helper_test.go:1037: Polling run "run-QRfJ7f5MdENu3mTD" for status included in ["cost_estimated" "errored"] with deadline of 2022-11-10 14:40:18.011966 +1100 AEDT m=+191.918870751
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "plan_queued"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "planning"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "planning"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "planning"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "post_plan_running"
    helper_test.go:1043: ...
    helper_test.go:1064: Reading run "run-QRfJ7f5MdENu3mTD"
    helper_test.go:1049: Run "run-QRfJ7f5MdENu3mTD" had status "cost_estimated"
    --- PASS: TestTaskStageOverride_Beta/when_the_policy_passed (51.83s)
PASS
ok     github.com/hashicorp/go-tfe     119.055s
?       github.com/hashicorp/go-tfe/examples/organizations      [no test files]
?       github.com/hashicorp/go-tfe/examples/workspaces [no test files]
?       github.com/hashicorp/go-tfe/mocks       [no test files]
```

### CODE REVIEW

File: run.go [hunk 1]
```diff
@@ -78,6 +78,7 @@ const (
 	RunPrePlanCompleted   RunStatus = "pre_plan_completed"
 	RunPrePlanRunning     RunStatus = "pre_plan_running"
 	RunQueuing            RunStatus = "queuing"
+	RunAwaitingDecision   RunStatus = "post_plan_awaiting_decision"
```
Comment (glennsarti) [inline run.go:81]: The name should be prefixed with the stage as overrides _can_ occur at any stage. https://github.com/hashicorp/atlas/blob/main/app/models/run.rb#L36-L41
```suggestion
	RunPostPlanAwaitingDecision   RunStatus = "post_plan_awaiting_decision"
```

File: task_stages.go [hunk 2]
```diff
@@ -15,8 +15,12 @@ type TaskStages interface {
 	// Read a task stage by ID
 	Read(ctx context.Context, taskStageID string, options *TaskStageReadOptions) (*TaskStage, error)
 
-	// List all task stages for a given rrun
+	// List all task stages for a given run
 	List(ctx context.Context, runID string, options *TaskStageListOptions) (*TaskStageList, error)
+
+	// **Note: This function is still in BETA and subject to change.**
+	// Override a task stage for a given run
+	Override(ctx context.Context, taskStageID string) (*TaskStage, error)
```
Comment (glennsarti) [inline task_stages.go:23]: I think this is missing the option to add a comment to the override. e.g. https://github.com/hashicorp/go-tfe/blob/main/run.go#L43

``` go
// TaskStageOverrideOptions represents the options for overriding a TaskStage.
type TaskStageOverrideOptions struct {
	// An optional explanation for why the stage was overridden
	Comment *string `json:"comment,omitempty"`
}
```

```suggestion
	Override(ctx context.Context, taskStageID string, options TaskStageOverrideOptions) (*TaskStage, error)
```

**Comment by mrinalirao** [comment]
Added test for passing optional comment into the Override API:

$ envchain local go test -run TestTaskStageOverride_Beta -v ./... -tags=integration

```
=== RUN   TestTaskStageOverride_Beta
--- PASS: TestTaskStageOverride_Beta (201.95s)
=== RUN   TestTaskStageOverride_Beta/when_the_policy_failed
    helper_test.go:1038: Polling run "run-JRmxhdQ9MS4jaz56" for status included in ["post_plan_awaiting_decision" "errored"] with deadline of 2022-11-11 14:19:36.167376 +1100 AEDT m=+151.096747460
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "plan_queued"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-JRmxhdQ9MS4jaz56"
    helper_test.go:1050: Run "run-JRmxhdQ9MS4jaz56" had status "post_plan_awaiting_decision"
    --- PASS: TestTaskStageOverride_Beta/when_the_policy_failed (82.70s)
=== RUN   TestTaskStageOverride_Beta/when_the_policy_failed_with_options
    helper_test.go:1038: Polling run "run-cn4CBHJ2nricvEb4" for status included in ["post_plan_awaiting_decision" "errored"] with deadline of 2022-11-11 14:21:01.832324 +1100 AEDT m=+236.761400418
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "plan_queued"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-cn4CBHJ2nricvEb4"
    helper_test.go:1050: Run "run-cn4CBHJ2nricvEb4" had status "post_plan_awaiting_decision"
&{ts-HPyoDwZhyZknXz1B post_plan running {0001-01-01 00:00:00 +0000 UTC 2022-11-11 03:19:27 +0000 +0000 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC} 2022-11-11 03:19:01.522 +0000 UTC 2022-11-11 03:19:27.335 +0000 UTC 0x1400000fe60 0x140002341f0 0x1400056c900 [] [0x140002701e0]}
    --- PASS: TestTaskStageOverride_Beta/when_the_policy_failed_with_options (77.08s)
=== RUN   TestTaskStageOverride_Beta/when_the_policy_passed
    helper_test.go:1038: Polling run "run-NbLgbVuBbkbDmyBu" for status included in ["cost_estimated" "errored"] with deadline of 2022-11-11 14:22:09.554191 +1100 AEDT m=+304.483033335
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "plan_queued"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "planning"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "post_plan_running"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "cost_estimating"
    helper_test.go:1044: ...
    helper_test.go:1065: Reading run "run-NbLgbVuBbkbDmyBu"
    helper_test.go:1050: Run "run-NbLgbVuBbkbDmyBu" had status "cost_estimated"
    --- PASS: TestTaskStageOverride_Beta/when_the_policy_passed (41.56s)
PASS
```

**Review by glennsarti — DISMISSED** [review]

File: run.go [hunk 3]
```diff
@@ -53,31 +53,32 @@ type RunStatus string
...
+	RunPolicyChecked            RunStatus = "policy_checked"
+	RunPolicyChecking           RunStatus = "policy_checking"
+	RunPolicyOverride           RunStatus = "policy_override"
+	RunPolicySoftFailed         RunStatus = "policy_soft_failed"
+	RunPostPlanCompleted        RunStatus = "post_plan_completed"
+	RunPostPlanRunning          RunStatus = "post_plan_running"
+	RunPrePlanCompleted         RunStatus = "pre_plan_completed"
+	RunPrePlanRunning           RunStatus = "pre_plan_running"
+	RunQueuing                  RunStatus = "queuing"
+	RunPostPlanAwaitingDecision RunStatus = "post_plan_awaiting_decision"
```
Comment (brandonc) [inline run.go:81]: I think these are in alphabetical order 🧐

File: task_stages.go [hunk 4]
```diff
@@ -15,8 +15,12 @@ type TaskStages interface {
 	// Read a task stage by ID
 	Read(ctx context.Context, taskStageID string, options *TaskStageReadOptions) (*TaskStage, error)
 
-	// List all task stages for a given rrun
+	// List all task stages for a given run
 	List(ctx context.Context, runID string, options *TaskStageListOptions) (*TaskStageList, error)
+
+	// **Note: This function is still in BETA and subject to change.**
```
Comment (brandonc) [inline task_stages.go:21]: 🙏

File: task_stages.go [hunk 5]
```diff
@@ -33,19 +37,51 @@ const (
...
+// Permissions represents the permission types for overridding a task stage
+type Permissions struct {
+	CanOverridePolicy *bool `jsonapi:"attr,can-override-policy"`
+	CanOverrideTasks  *bool `jsonapi:"attr,can-override-tasks"`
+	CanOverride       *bool `jsonapi:"attr,can-override"`
+}
+
+// Actions represents a task stage actions
+type Actions struct {
+	IsOverridable *bool `jsonapi:"attr, overridable"`
```
Comment (brandonc) [inline task_stages.go:60]: I don't think this space is a problem but if you are already committing anything else, maybe just delete it so my eye doesn't twitch

```suggestion
	IsOverridable *bool `jsonapi:"attr,overridable"`
```

Comment (brandonc) [reply task_stages.go:60]: Looking at the serializer, it looks like the property is "is-overridable"

can you double check that?

**Review by brandonc — COMMENTED** [review]
I just have the one question about is-overridable attribute.

What is your expectation about the release timing of this, given that is it still feature flagged in the API?

File: task_stages.go (hunk 5 above)
Comment (mrinalirao) [reply task_stages.go]: Good catch :) Fixed.

**Comment by mrinalirao** [comment]
> What is your expectation about the release timing of this, given that is it still feature flagged in the API?

We would want to release this now since we want to use this version of the client for adding OPA support to both the provider and the cli. We are targeting to get the provider out before year end holidays and the cli before the end of Jan. Once OPA goes to GA, we will come back to this to remove the beta notes.

**Review by brandonc — APPROVED** [review]

**Pull request merged**